### DISCUSSION TITLE
Entity Decoder - Remove EitherT

### DISCUSSION THREAD

**Discussion by diesalbla** [http4s/http4s#6875] [Show and tell]
The PR https://github.com/http4s/http4s/pull/6846 carries the changes to the `main` branch needed to remove the use of `EitherT` from 

The PR changes the external API of `http4s` in these two ways.

- The type alias `DecodeResult` changes from referring to the `EitherT` monad transformer to an `Either` of a failure or the result type. This makes it similar to the nearby `ParseResult` alias, or to the `Decoder.Result` alias in [circe](https://github.com/circe/circe/blob/series/0.14.x/modules/core/shared/src/main/scala/io/circe/Decoder.scala#L460).

```Scala
type DecodeResult[F[_], T] = EitherT[F, DecodeFailure, T]
type DecodeResult[T] = Either[DecodeFailure, T]
```

The interface for an `EntityDecoder` changes from returning the previous `EitherT` alias, to the unwrapped `F[Either[DecodeFailure, A]]`.

```Scala
trait EntityDecoder[F[_], T] {
  def decode(m: Media[F], strict: Boolean): DecodeResult[F, T]
  def decode(m: Media[F], strict: Boolean): F[DecodeResult[T]]
}
```

The motivation for this change is to remove the use of the `EitherT` in this part of the public API. 

From what I have seen in the PR, the only disadvantage is loosing some of the methods from the `EitherT` class, but these are not widely used and their non-transformed replacement do not take that many lines in comparison. 

### Discussion or motivations

An `EitherT[F[_], L, R]` is a wrapper around an `F[Either[L, R]]`. Choosing to use one or not comes down to whether it helps to understanding, makes it easier to use the value, and fits the context where it's used. Since no one can understand the wrapper without its contents, the current way is at least one step  more difficult. The only way a transformer help is when used in MTL-style code, which generalizes over other notions of monad. However, this is not commonly used for HTTP routes, and when used it is better left as an opt-in than an opt-out. 

The purpose of an `EntityDecoder` is to decode the entities that accompany a request in a server, or a response in a server. In both the server and client definitions, there are other forms of error raising and error handling that are not done through `EitherT`, so this wrapper does not fit in the context where it is often used.

**Comment by rossabaker** [comment]
All breaking changes should be run through a cost-benefit analysis.

1. **Does it improve correctness?**  No effect.
2. **Does it improve ergonomics?** Perhaps slightly.  It's more convenient if `decode` is unlikely to be composed with other `F` and `Either`.  If both monads are likely to be present, it's worse than `EitherT`.  In practice, I don't think it matters much: few people are calling this interface directly, but instead via `Message#as` or `Client#expect`, which eliminate the `Either` into the untyped error channel.
3. **Does it improve performance?**  Trivially, but when `F` is `IO` representing network calls, it probably doesn't matter.  There are a lot of frankly dishonest benchmarks from other camps that have poisoned this discourse.

In a vacuum, I see a small benefit.

4. **Can we do a deprecation cycle?** Yes, but only if we gave the new method a new name.  This proposed approach is a hard break.
5. **Can we scalafix to fix the breakage or deprecations?** Likely, but there's no PoC.
6. **How sweeping is the change?** Mitigated by the use of `.as` and `.expect`.

The change could be made cheaper which would make this more appealing, but it's one more hurdle to the adoption of 1.0.  At $WORK, many projects are stuck on 0.21, and some further behind that, because breaking changes are really expensive, and the more an upgrade costs, the less likely it is to ever happen.  This increases developer drag: I spent part of my day off yesterday preparing a security fix for _four versions_.  I think the bar here is not "makes things a little nicer," but rather, "is clearly worth the hassle of developer attention to upgrade and maintain."  I don't see that the benefits clearly outweigh the costs here.

If we were to make a grander change, like EntityCodecs as _data_ (i.e., explicit, like Skunk) rather than as a type class instance (i.e., implicit, like Circe), then I could imagine this being part of it: that change is much more expensive _and_ beneficial.  But that proposal is a whole other discussion.

**Reply by diesalbla** [reply]
> 3. Does it improve performance? Trivially, but when F is IO representing network calls, it probably doesn't matter. There are a lot of frankly dishonest benchmarks from other camps that have poisoned this discourse.

Benchmarks are of course an approximation easy to steer. It would be nice to have a source of real-world observations. For instance, if one knew a company running dozens of micro-services using clients and servers in `http4s`, one could try to profile their production systems to measure how much those `EitherT` objects really weight on the memory allocation costs.

> 4. Can we do a deprecation cycle? Yes, but only if we gave the new method a new name. This proposed approach is a hard break.

A deprecation cycle may be more amenable, it would be easy to keep the existing method as deprecated and move the logic to a new method.

The cost-benefit analysis can also account for making the library amenable by reducing the knowledge require to use it, or just read it, by getting rid of an unhelpful notion. The stability argument appeal to current users, the ergonomics / learning curve can help potential new users. 

> If we were to make a grander change [...], then I could imagine this being part of it: that change is much more expensive and beneficial. But that proposal is a whole other discussion.

It could be part of several changes together with the change of `HttpRoutes` away from `Kleisli` / `OptionT` of https://github.com/http4s/http4s/pull/6769/files, and other changes such as https://github.com/http4s/http4s/pull/6738, oriented to reduce the use of monad transformers in the public api.

**Reply by valencik** [reply]
@rossabaker thanks for the detailed thoughts here.
I think this perspective is really helpful and quite clearly stated:

> I think the bar here is not "makes things a little nicer," but rather, "is clearly worth the hassle of developer attention to upgrade and maintain."

**Comment by rossabaker** [comment]
> If we were to make a grander change, like EntityCodecs as data 

Actually, both this proposal and that are part of my grander idea: to finally support content negotiation.  I will try to find time for a sketch of that.
